Document ZFS macOS setup. (#42)
ocurrent/macos-infra: README-rsync.md

<!-- TOC -->

- [OCluster/OBuilder macOS Infrastructure](#oclusterobuilder-macos-infrastructure)
- [The Big Picture](#the-big-picture)
- [The Implementation](#the-implementation)
    - [Overview](#overview)
    - [Snapshotting Filesystem](#snapshotting-filesystem)
    - [Global System Dependencies](#global-system-dependencies)
    - [Tricking Homebrew with FUSE](#tricking-homebrew-with-fuse)
    - [Rsyncing to Home](#rsyncing-to-home)
    - [Docker-esque Base Images](#docker-esque-base-images)
- [OCluster macOS Worker](#ocluster-macos-worker)
    - [Building a macOS Base Directory](#building-a-macos-base-directory)
- [Ansible Playbook](#ansible-playbook)
- [Starting and Stopping](#start-stop)
- [Current Deployment and Future Steps](#current-deployment-and-future-steps)
- [Thanks](#thanks)

<!-- /TOC -->

## The Big Picture

[OCurrent](https://github.com/ocurrent/ocurrent) is an OCaml library for building incremental pipelines. Pipelines can be thought of as directed graphs with each node being some kind of job that feeds its results into the next node. For an example, have a look at [the OCaml deploy pipeline](https://deploy.ci.ocaml.org/).

> An integral part to most of the [OCaml pipelines](https://github.com/ocurrent/overview) is something called [OCluster](https://github.com/ocurrent/ocluster) and [OBuilder](https://github.com/ocurrent/obuilder). OCluster is a library and collection of binaries for managing an infrastructure for building and scheduling jobs. These jobs could be Docker jobs or OBuilder jobs. OBuilder is quite similar to `docker build` but written in OCaml and using just a notion of *an execution environment* (on Linux [`runc`](https://github.com/opencontainers/runc)) and a *snapshotting filesystem* ([Btrfs](https://btrfs.wiki.kernel.org/index.php/Main_Page), [ZFS](https://openzfs.org/wiki/Main_Page), or an inefficient but convenient copying backend using `rsync`).

Reviewer: Should `runc` be in monospace or like this: runc ?
Same for `rsync`. I've seen it written Rsync when short for Remote Sync and rsync in Wikipedia. Or is it a command `rsync`. It seems throughout this document it's used sometimes as a command and sometime as the utility itself.

Reviewer 2: `runc` and `rsync` are both system commands.
* https://github.com/opencontainers/runc
* https://rsync.samba.org or https://github.com/WayneD/rsync

> An integral part to most of the [OCaml pipelines](https://github.com/ocurrent/overview) is something called [OCluster](https://github.com/ocurrent/ocluster) and [OBuilder](https://github.com/ocurrent/obuilder). OCluster is a library and collection of binaries for managing an infrastructure for building and scheduling jobs. These jobs could be Docker jobs or OBuilder jobs. OBuilder is quite similar to `docker build` but written in OCaml and using just a notion of *an execution environment* (on Linux [`runc`](https://github.com/opencontainers/runc)) and a *snapshotting filesystem* ([Btrfs](https://btrfs.wiki.kernel.org/index.php/Main_Page), [ZFS](https://openzfs.org/wiki/Main_Page), or an inefficient but convenient copying backend using `rsync`).

Reviewer: ```suggestion
An integral part to most of the [OCaml pipelines](https://github.com/ocurrent/overview) is something called [OCluster](https://github.com/ocurrent/ocluster) and [OBuilder](https://github.com/ocurrent/obuilder). OCluster is a library and collection of binaries for managing an infrastructure that builds and schedules jobs. These jobs could be Docker jobs or OBuilder jobs. OBuilder is quite similar to `docker build` but written in OCaml and using just a notion of *an execution environment* (on Linux [`runc`](https://github.com/opencontainers/runc)) and a *snapshotting filesystem* ([Btrfs](https://btrfs.wiki.kernel.org/index.php/Main_Page), [ZFS](https://openzfs.org/wiki/Main_Page), or an inefficient but convenient copying backend using `rsync`).
```

The big picture, and focus of this document, is adding support for building macOS OBuilder jobs. It's important to note the narrow scope of the project. For this iteration, we just need to build [opam](https://opam.ocaml.org) packages.

## The Implementation

### Overview

The main problem we are tackling is that macOS has *no notion or support of native containers*. You can have multiple users on a single machine, each with their own home directory and ability to execute commands and a way of identifying who they are (their `uid`). This is how the macOS implementation of OBuilder's `SANDBOX` is built. Individual users can be created for each build, and in the future `sandbox-exec` can help to restrict what they have access to.

Although not part of the current implementation, the idea is to have multiple users building in parallel. For now, it is restricted to just one.

### Snapshotting Filesystem

> On macOS, there is a [port of ZFS](https://openzfsonosx.org/) that, at the time, works quite well but not perfectly. Many hours were lost debugging what ended up being small bugs in ZFS. This is what inspired adding a very portable and reliable, but slow and memory-inefficient [`rsync` store backend](https://github.com/ocurrent/obuilder/pull/88). This is what is currently used in the macOS implementation, which so far has been very reliable, and once the caches are hot, the slow down isn't too noticeable.

Reviewer: ```suggestion
On macOS, there is a [port of ZFS](https://openzfsonosx.org/) that, at the time, worked quite well but not perfectly. Many hours were lost debugging what ended up being small bugs in ZFS. This is what inspired adding a very portable and reliable, but slow and memory-inefficient [`rsync` store backend](https://github.com/ocurrent/obuilder/pull/88). This is what is currently used in the macOS implementation, which so far has been very reliable, and once the caches are hot, the slow down isn't too noticeable.
```

Should this be "worked" since were talking past tense "at the time"?

### Global System Dependencies

Opam, the OCaml package manager, is happy to have multiple opam roots on a system typically in the home directory of the user (`~/.opam`). This is great as running something like `sudo -u macos-builder-705 -i /bin/bash -c 'opam install irmin'` will use `macos-builder-705`'s home directory to build `irmin`. This provides an illusion of isolation similar to `runc`.

> Homebrew, a macOS system package manager, is not so willing to be installed anywhere. [You can do it](https://docs.brew.sh/Installation#untar-anywhere), but as they say:

Reviewer: ```suggestion
Homebrew, a macOS package manager, is not so willing to be installed anywhere. [You can do it](https://docs.brew.sh/Installation#untar-anywhere), but as they say:
```
Tighter, if "system" will be understood without it there

> ...do yourself a favour and install to `/usr/local` on macOS Intel, `/opt/homebrew` on macOS ARM ... Pick another prefix at your peril!

### Tricking Homebrew with FUSE

We want to have multiple Homebrew installations that all believe they are installed in `/usr/local`. Without this, Homebrew can't use "bottles," which are pre-built binaries making builds much faster.

> The proposed solution is to mount a FUSE (filesystem in userspace) filesystem onto `/usr/local` which intercepts calls and redirects them based on who the calling user is. Since macOS Catalina, this requires [System Integrity Protection (SIP) to be disabled](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection).

Reviewer: ```suggestion
The proposed solution is to mount a FUSE (filesystem in userspace) onto `/usr/local` which intercepts calls and redirects them based on who the calling user is. Since macOS Catalina, this requires [System Integrity Protection (SIP) to be disabled](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection).
```
Since the F in FUSE stands for filesystem, does it need to specify a filesystem as well?

<p align="center">
<img style="width:40%" src="./docs/fuse.svg" alt="A diagram showing the mapping from /usr/local to the calling user's home directory."/>
</p>

FUSE luckily supplies all calls with a `fuse_context`, which provides the `uid` of the calling user. We can use this to redirect to the correct home directory. This is what the [`obuilder-fs`](https://github.com/patricoferris/obuilder-fs) filesystem does. The implementation ensures that it is mounted, and all calls to `/usr/local` are redirected. This does have some cost, but in practice, most packages make use of a few core system dependencies.

### Rsyncing to Home

> With the scheme described above, each user's home directory is easy to find, as it is based on the `uid`. One problem though is that for each build step (`run`, `copy`...) we take a snapshot so we can later restore from it. This means we need to sync the home directory with the snapshot. This is also important for packages that are not relocatable (at the time of writing `ocamlfind` is an example).

Reviewer: ```suggestion
With the scheme described above, each user's home directory is easy to find, as it is based on the UID. One problem though is that for each build step (`run`, `copy`...) we take a snapshot so we can later restore from it. This means we need to sync the home directory with the snapshot. This is also important for packages that are not relocatable (at the time of writing OCamlFind is an example).
```
Should these be in monospace or be the acronym and proper name, respectively?

The workaround is to `rsync` the snapshot from the store to the user's home directory every time we execute a step. This is exactly what happens in this [code in the macOS sandbox implementation](https://github.com/patricoferris/obuilder/blob/8c3f200b519ad14a5f70787e42f59ec7db229d3c/lib/sandbox.macos.ml#L61).

### Docker-esque Base Images

OBuilder spec files start with a `(from ...)` stage. This identifies the image that should be used as a basis for the build. In OCaml-related projects, this tends to be one of the [Docker base images](https://base-images.ocamllabs.io/).

In order to minimise the amount of additional logic that would be needed for macOS inside things like [ocaml-ci](https://ci.ocamllabs.io/), [opam-repo-ci](https://github.com/ocurrent/opam-repo-ci), and [opam-health-check](http://check.ocamllabs.io/), it makes sense for the macOS version to be as similar as possible to the Docker base images.

This includes:

 - Having both `opam.2.0.X` and `opam.2.1.X` installed and ready to use by sym-linking to `/usr/local/bin/opam` (a.k.a. `~/local/bin/opam`).

>  - Being clever (thanks @kit-ty-kate) with the `.bash_profile` script to reuse the name of the base image (i.e., `(from "macos-homebrew-ocaml-4.11")`) to setup things like the path to the system compiler. This means the OBuilder specs don't need to worry about the OCluster worker implementation details.

Reviewer: ```suggestion
 - Being clever (thanks @kit-ty-kate) with the `.bash_profile` script that reuses the base image name (i.e., `(from "macos-homebrew-ocaml-4.11")`) to setup things like the path to the system compiler. This means that the OBuilder specs don't need to worry about the OCluster worker implementation details.
```

We still use system compilers installed on the worker machine. The OBuilder implementation always sources a `.obuilder_profile.sh` file before building. There we can ensure the correct `PATH` is set. To easily install system compilers (for different versions of the compiler), there is [the opam-sysinstall plugin](https://github.com/patricoferris/opam-sysinstall).

To build a new base directory you can run:

```
sudo dune exec -- ./main.exe --ocaml-version=4.14.0 --rsync=/Volumes/rsync --uid=705 --fallback=/Users/administrator/lib --scoreboard=/Users/administrator/scoreboard --verbosity=info
```

The final build will be in `/Users/mac705` (or in the final snapshot in the `rsync` store). If you want to use the `User_temp` copying `FETCHER` then you can

```
sudo rsync -aHq /Users/mac705/ /Users/macos-homebrew-ocaml-4.14
```

> And use this from stage in your spec file `(from macos-homebrew-ocaml-4.14)`. Note that the OCluster PR still uses the Docker fetcher in the PR, but not on the machines. The machines have the following small patch that hopefully in the future can be removed.

Reviewer: ```suggestion
And use this from the stage in your spec file `(from macos-homebrew-ocaml-4.14)`. Note that the OCluster PR still uses the Docker fetcher in the PR, but not on the machines. The machines have the following small patch that hopefully in the future can be removed.
```

Should this be "the stage" or "a stage"?

- Energy Saver \ Start up automatically after power failure
- Install Apple Developer Command Line Tools
- Turn off SIP by entering Recovery Mode (Intel: Command-R; M1: hold power button) `csrutil disable`
- Install [macFUSE](https://osxfuse.github.io), which requires approval via System Preferences and a reboot of the system.
  Use `brew install --cask macfuse` which will require a Homebrew install.
- Install [Docker Desktop for Mac](https://docs.docker.com/desktop/mac/install/)
- Set Docker to automatically start at sign in
- Add the capability file for the scheduler pool you want to join to `./secrets/pool-macos-[arch].cap` (creating the `secrets` directory in the project root). For more info on the scheduler, see [ocurrent/ocluster](https://github.com/ocurrent/ocluster).
- Add your SSH key to the `~/.ssh/authorized_keys` and update your `~/.ssh/config` so that you can SSH to the Mac without prompting for a username:

```
Host mac-mon-*
	User administrator
```

Run the playbook as follows.  I have used `--limit` to target a single worker.

```sh=
ansible-playbook -i hosts --limit i7-worker-01 playbook.yml
```

## Starting and Stopping

> `ocluster-worker` is run via LaunchAgent.

Reviewer: ```suggestion
`ocluster-worker` is run via LaunchAgent.
```
Should this be OCluster-Worker or the command `ocluster-worker`. Seems like the latter. It was written Ocluster-worker, but since it's being run via LaunchAgent, I figured it was a comman: `ocluster-worker`.
